### DISCUSSION TITLE
Development update 3 April - Arc Overhang fluid recursion implementation progress

### DISCUSSION THREAD

**Discussion by rvmn** [rvmn/SuperPleccer#10] [Announcements]
Hi all!  
A little update on the progress of the new Pleccer beta coming in the coming weeks, it's coming closer each day. 
Focus now lies on fixing the last problems with the mixing extruder support the color picker but there is also progress to show on the arc overhang support of Pleccer!

A rewrite of the Arc Overhang algorithm has been worked out that adds the inclusion of ´fluid´-recursion.
The recursion is called fluid because it prints full fluidly; meaning that no z-lifts or travel moves are needed (except needed side-to-side travel for letting the line cool). There should be no sagging or blobbing happening this way.
The result does become a bit different from the regular recursive arcs.

To show it a few pictures may say the most.

![Schermafdruk van 2023-04-03 01-53-19](https://user-images.githubusercontent.com/8228964/229386409-22163422-9035-4a28-b6a4-6ae7514a2038.png) 25%
![Schermafdruk van 2023-04-03 01-54-10](https://user-images.githubusercontent.com/8228964/229386415-7d311cec-6772-47fa-8e8e-a7361dcf3dcd.png) 50%
![Schermafdruk van 2023-04-03 01-54-42](https://user-images.githubusercontent.com/8228964/229386426-53b2d51c-8d64-45c7-ad55-f5369d23e793.png) 75%
![Schermafdruk van 2023-04-03 01-55-33](https://user-images.githubusercontent.com/8228964/229386433-ae1428df-e25a-4140-aef9-f295ff270c6b.png) 100%

Right now the only parameter available is a percentage of 0 - 100% (images are of 25%, 50%, 75% and 100% resp.), but other parameters will be added in the coming week.

As you can see there is quite a big random factor and there are still some issues. The rsndom factor is hard to defeat as the algorithm calculates everything as cheaply as possible while doing the recursing and things are meant to be fluid anyhow. Next main improvement to use will be adding Arachne lines to make sure the extrusion width is adaptive so the lines adapt a bit and no holes will be in the layer. Also making the big corners less sharp using corner limits. Expect these improvements in the beta coming in the next weeks.

FYI no arcs were harmed in the making of this update!

Greets,
Pleccer
![Schermafdruk van 2023-04-02 21-16-40](https://user-images.githubusercontent.com/8228964/229386340-82b1a458-3a4a-4f83-95a1-0c87ee62f615.png)
![Schermafdruk van 2023-04-03 02-28-33](https://user-images.githubusercontent.com/8228964/229388479-c10405b3-145c-4294-b684-d5b4c9596ece.png)

**Comment by stmcculloch** [comment]
@rvmn Very interesting idea. Is it supposed to simulate a liquid filling the available space? 

Have you thought about incorporating a "flood fill" algorithm instead of using arcs? These toolpaths seem promising: https://github.com/stmcculloch/arc-overhang/issues/12

General idea of floodfill: 
![image](https://user-images.githubusercontent.com/9290784/229439449-4deca334-a321-42d9-bdc4-03267a09d6cc.png)

**Reply by rvmn** [reply]
@stmcculloch yes exactly, fluid in the sense of adapting to the limits surrounded by. 

flood fill is alright, i just see the start of it not being as reliable as arcs are. 

the first lines from the starting edge in your example above are just simple 90deg rectilinear lines, that is what has been used for overhangs by all  the 'witty' printerers. you basically have to print very slow and hope there is enough accidental upward moving turbulence caused by the part cooling fan to not have it sag ;)   
[this](https://help.prusa3d.com/article/poor-bridging_1802) will probably happen.

btw, the arcs by themselves become flood fill at a later stage, the later stage of the flood fill is not different from the lines in above arc-infill pictures.

**Reply by rvmn** [reply]
New update on exploring arc recursion
![Schermafdruk van 2023-04-17 02-24-22](https://user-images.githubusercontent.com/8228964/232352537-f30bb0cd-f28d-47c4-acf3-c3d1e6bda10f.png)
Its a lot cleaner now already and almost without gaps. Getting Arachne at work for it was a bit of a impossibillity as that only works for concentric loops like walls and concentric infill, and not for seperate lines.

Its proven to be pretty hard to get right, because it is having to do a multi-round loop arc-by-arc to recreate the next offset outward, check the outcome, modify if needed, then rebuild the loop offset, check again and so on, so it is a pretty costly algorithm. And getting a parameter too low means bad precision and gaps as result.

**Reply by rvmn** [reply]
Update: 
The major work regarding bridging and overhangs has been finished, now its only over to make the right settings available.
Nice added addition is an option to turn overhangs to bridges when a settable distance is reached (30 mm defaulting, otherwise arcing takes a lot of time).
The bridging also has a setting of how far the bridging can be reliably done, max bridge distance (20mm default). As always a small image of the result:
![Schermafdruk van 2023-04-19 22-31-41](https://user-images.githubusercontent.com/8228964/233194542-f54704c4-8a6b-4d21-831c-bae40dcb6d96.png)
As you can see, in this image, quote some supports, but this is on medium support setting, it can be set to low support also.
![Schermafdruk van 2023-04-19 22-42-05](https://user-images.githubusercontent.com/8228964/233195245-6a157d5b-be30-4817-a4c3-44eaba9071a6.png)

**Comment by rvmn** [comment]
The, internally called ´arc_diffraction´, recursioning is no ´real´ recursioning but should offer quite some of it's effect also.
![Schermafdruk van 2023-04-19 22-49-54](https://user-images.githubusercontent.com/8228964/233196913-6d1c59f2-0298-494f-ae13-bc9934b82972.png)
Right now there is only an overall strength setting (0-100), showing above is 50%; the middle setting, but internally it has periodicity, amplitude and slit width:
![hq720](https://user-images.githubusercontent.com/8228964/233197762-e4ab83d1-d757-463f-a72f-a5e407d258d8.jpg)

![Schermafdruk van 2023-04-19 23-07-22](https://user-images.githubusercontent.com/8228964/233200332-e03662b1-7817-43a4-8c63-2f47ff9c5d33.png)

[Schermopname van 24-04-23 21:20:38.webm](https://user-images.githubusercontent.com/8228964/234096476-4decff93-aaf7-485c-9ff8-f51352e0b7a0.webm)

**Comment by rvmn** [comment]
[Schermopname van 20-04-23 09:57:52.webm](https://user-images.githubusercontent.com/8228964/233300059-1ee03178-a918-4cea-b146-49e755e68843.webm)

[Schermopname van 20-04-23 11:26:03.webm](https://user-images.githubusercontent.com/8228964/233322620-46113c89-3a11-44a2-ac78-b0787a624d5f.webm)

**Comment by rvmn** [comment]
bridging 

[Schermopname van 20-04-23 11:37:18.webm](https://user-images.githubusercontent.com/8228964/233325652-ccf4f019-60f1-40ce-a68b-c65d12949419.webm)

[Schermopname van 20-04-23 11:37:54.webm](https://user-images.githubusercontent.com/8228964/233325691-df83a6eb-a806-4252-b5fa-9feba6952aa8.webm)

**Comment by stmcculloch** [comment]
Really interesting developments! Any test prints?

**Reply by rvmn** [reply]
yes, but it hasn't been anything really worth it so far, the new arc starting points, even though they are started as part in the previous line, still sag a little bit, not much, just more than i expected them to. 
as printing is just a hobby, i have to find time to print some more test files, just finished with upgrading my 2 printers and they should run fine for a while now (have a flsun sr delta and a tronxy corexy), they now print faster and have much better cooling, even though for testing i mightr turn that down a bit to get accurate 'stock 3d printer' results.
also, as last small thing, not an anouncement yet, but still, might be interesting:  searching stuff on 5-axis printing i found [curvislicer](https://github.com/mfx-inria/curvislicer) which doesnt even need anything special, and adds non-planar slicing. and that uses iceSL slicer, which is also really interesting. we are all bound to prusa being the 'best FOSS slicer' but iceSL is way more modernly programmed with features that make programming on it fun, where prusa is old-style 'try-not-to-create-more-spaghetti' type programming, i like that is CPP, but it is archaically programmed tbh. 
Features that really make iceSL stand strong are among others:  
  - GLSL and image based infill, GLSL (shading language in C) is used for many things, like shading (which is putting light on 3d objects) and uses a very close to math language. See more [here](https://github.com/shapeforge/icesl-infillers)
  - very nice, really snug, support which imho beats tree support or at least doesnt/hardly uses more material. [More info](https://icesl.loria.fr/pages/features/advanced-bridge-support/)
 [Full feature list](https://icesl.loria.fr/features/)
 
 In context: i tried bambu/orcaslicer, but they have become very heavy and are too slow on my 2020 surface laptop, bambu hasnt thought about performance when programming stuff in. Prusaslicer is programmed to perform but doesnt have very good printing results, it is missing logic in the code, seems to be made more by programmers and ideaists than hardcore 3d printerers. Superslicer is a working upgrade from prusa, much better printing results and actually has a lot of good working print features. it just doesn't have good support printing feature/ lacks adaptive layer height and some special options. IceSL has those all, just doesn't have a name/status yet.

SO concluding: i will/am continuing with SuperPleccer as SuperSlicer based fork. And i am going to spend some time understanding iceSL code
